### ISSUE TITLE
ld.so with non-contiguous segments breaks startup

### ISSUE THREAD

**Issue by stephenrkell** [stephenrkell/liballocs#47]
If the ld.so has a hole in, it will be mapped by the kernel with that hole existing. This leads the mmap allocator to build not a single mapping sequence for it but two. This in turn foils the static file allocator which expects a single mapping sequence for the whole file.

My first approach was to plug the hole by mapping it PROT_NONE. However, it may be too late -- sometimes a TLS area, or other random stuff, has been placed in the hole.

Currently I'm seeing this on the CircleCI machine using Ubuntu 18.04.

We may need to plug the gap super-early, or else detect the problematic ld.so build and request the user to patch it (recalling that in the long run, we want to *be* the ld.so).

Note that non-ld.so binaries are fine because they are mapped by the ld.so, which does not create holes... it uses PROT_NONE.

**Comment by stephenrkell** [comment]
It gets worse. After hacking an infinite loop as early as I could into the startup code, then inspecting /proc/pid/maps, we see...

```
7f3e0e18c000-7f3e0e1b3000 r-xp 00000000 00:f0 29                         /lib/x86_64-linux-gnu/ld-2.27.so
7f3e0e39b000-7f3e0e3a6000 rw-p 00000000 00:00 0 
7f3e0e3ac000-7f3e0e3af000 r-xp 00000000 00:f0 83376                      /root/project/build/debug/liballocs_dummyweaks.so
7f3e0e3af000-7f3e0e3b0000 r--p 00002000 00:f0 83376                      /root/project/build/debug/liballocs_dummyweaks.so
7f3e0e3b0000-7f3e0e3b1000 rw-p 00003000 00:f0 83376                      /root/project/build/debug/liballocs_dummyweaks.so
7f3e0e3b1000-7f3e0e3b3000 rw-p 00000000 00:00 0 
7f3e0e3b3000-7f3e0e3b4000 r--p 00027000 00:f0 29                         /lib/x86_64-linux-gnu/ld-2.27.so
7f3e0e3b4000-7f3e0e3b5000 rw-p 00028000 00:f0 29                         /lib/x86_64-linux-gnu/ld-2.27.so
7f3e0e3b5000-7f3e0e3b6000 rw-p 00000000 00:00 0 
```
i.e. the ld.so has already mapped a library in the hole, not just TLS stuff.

If we can't get control early enough, one workaround is to do chain loading of the ld.so. We could create a simple "no holes" ld.so that delegates to the real ld.so, and hey presto.

The right thing to do is to switch more directly to the libgerald approach of being a custom ld.so. To understand the quickest path to something workable there, I need to find out more about how the ld.so interacts with preloaded libraries -- does it come first in the link order? I'll do that on a 'with-libgerald' branch soon, but the chain loader is fun enough to be worth hacking up for now.

**Comment by stephenrkell** [comment]
Just pushed an attempted fix to this, using the chain-loading approach.

**Comment by stephenrkell** [comment]
With that commit and a few smaller fixes (latest: e158425), this is now fixed.
